### PR TITLE
Fixes and improvements

### PR DESCRIPTION
**Pull request by ooddaa**
Various things that I stumbled into while paper-trading BAG option contracts

### CODE REVIEW

File: lib/ib_ex/client/messages/base.ex [hunk 1]
```diff
@@ -22,7 +22,7 @@ defmodule IbEx.Client.Messages.Base do
   end
 
   @spec message_id_from_fields(any()) :: {:ok, String.t()} | :error
-  def message_id_from_fields(fields) when is_list(fields) and length(fields) > 2 do
+  def message_id_from_fields(fields) when is_list(fields) and length(fields) > 1 do
```
Comment (ooddaa) [inline lib/ib_ex/client/messages/base.ex:25]: this to allow processing of `OptionChainEnd` message that only has two fields

File: lib/ib_ex/client/types/contract.ex [hunk 2]
```diff
@@ -99,15 +99,6 @@ defmodule IbEx.Client.Types.Contract do
...
   def new(attrs) when is_map(attrs) do
-    attrs =
-      case attrs.security_type == "BAG" do
-        true ->
-          assign_params(attrs, :delta_neutral_contract, DeltaNeutral)
-
-        false ->
-          attrs
-      end
-
```
Comment (ooddaa) [inline lib/ib_ex/client/types/contract.ex:110]: this was a mistake

File: lib/ib_ex/client/types/order/delta_neutral.ex [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  defstruct order_type: nil,
+            aux_price: nil,
+            conid: 0,
+            settling_firm: nil,
+            clearing_account: nil,
+            clearing_intent: nil,
+            open_close: nil,
+            short_sale: false,
+            short_sale_slot: 0,
+            designated_location: nil
```
Comment (ooddaa) [inline lib/ib_ex/client/types/order/delta_neutral.ex:11]: this was previously part of `volatility_order_params` but shouldn't have

File: lib/ib_ex/client/messages/market_data/request_cancel_data.ex [hunk 4]
```diff
@@ -19,10 +19,10 @@ defmodule IbEx.Client.Messages.MarketData.RequestCancelData do
   alias IbEx.Client.Messages.Requests
   alias IbEx.Client.Protocols.Subscribable
 
-  @spec new() :: {:ok, t()} | {:error, :not_implemented}
-  def new() do
+  @spec new(non_neg_integer()) :: {:ok, t()} | {:error, :not_implemented}
+  def new(request_id) do
```
Comment (CamonZ) [inline lib/ib_ex/client/messages/market_data/request_cancel_data.ex:23]: Why would you need to pass the `request_id`, arguably the same process that requested the data initially is now asking for it to be canceled.

File: test/ib_ex/client/messages/market_data/request_cancel_data_test.exs [hunk 5]
```diff
@@ -5,37 +5,39 @@ defmodule IbEx.Client.Messages.MarketData.RequestCancelDataTest do
   alias IbEx.Client.Protocols.Subscribable
   alias IbEx.Client.Subscriptions
 
+  @request_id 1000
   describe "new/1" do
     test "creates a RequestCancelData struct" do
-      assert {:ok, msg} = RequestCancelData.new()
+      assert {:ok, msg} = RequestCancelData.new(@request_id)
```
Comment (CamonZ) [inline test/ib_ex/client/messages/market_data/request_cancel_data_test.exs:11]: why these changes to pass the request_id?

File: test/ib_ex/client/messages/market_data/request_data_test.exs [hunk 6]
```diff
@@ -85,7 +85,7 @@ defmodule IbEx.Client.Messages.MarketData.RequestDataTest do
   end
 
   describe "Subscribable" do
-    test "subscribe/2 subscribes incoming messages with the msg's request id to the given pid" do
+    test "subscribe/3 subscribes incoming messages with the msg's request id to the given pid" do
```
Comment (CamonZ) [inline test/ib_ex/client/messages/market_data/request_data_test.exs:88]: nice catch on all these test titles

File: lib/ib_ex/client/types/order/delta_neutral.ex (hunk 3 above)
Comment (CamonZ) [reply lib/ib_ex/client/types/order/delta_neutral.ex:11]: how'd you figure it out?

Comment (ooddaa) [reply lib/ib_ex/client/types/order/delta_neutral.ex:11]: Mainly based off its naming as `deltaNeutral**`. Also the fields from `conid` to `designated_location` depend on `order.deltaNeutralOrderType` so makes sense to serialise those from `DeltaNeutral.serialize` struct.  

```python
                make_field_handle_empty(order.volatility),
                make_field_handle_empty(order.volatilityType),
                make_field(order.deltaNeutralOrderType),  # srv v28 and above
                make_field_handle_empty(order.deltaNeutralAuxPrice),
            ]  # srv v28 and above

            if (
                self.serverVersion() >= MIN_SERVER_VER_DELTA_NEUTRAL_CONID
                and order.deltaNeutralOrderType
            ):
                flds += [
                    make_field(order.deltaNeutralConId),
                    make_field(order.deltaNeutralSettlingFirm),
                    make_field(order.deltaNeutralClearingAccount),
                    make_field(order.deltaNeutralClearingIntent),
                ]

            if (
                self.serverVersion() >= MIN_SERVER_VER_DELTA_NEUTRAL_OPEN_CLOSE
                and order.deltaNeutralOrderType
            ):
                flds += [
                    make_field(order.deltaNeutralOpenClose),
                    make_field(order.deltaNeutralShortSale),
                    make_field(order.deltaNeutralShortSaleSlot),
                    make_field(order.deltaNeutralDesignatedLocation),
                ]
```

What do you think?

File: lib/ib_ex/client/messages/market_data/request_cancel_data.ex (hunk 4 above)
Comment (ooddaa) [reply lib/ib_ex/client/messages/market_data/request_cancel_data.ex:23]: This is the simplest way I could think of to make cancelling data subs actually work. Otherwise where would [`msg.request_id`](https://github.com/CamonZ/ib_ex/blob/072be72c1e9c132ece0b332da28b68c886ecad3e/lib/ib_ex/client/messages/market_data/request_cancel_data.ex#L34) come from? It has to match the `request_id` of the data subscription request being cancelled, and one pid can register be multiple data subs.

File: test/ib_ex/client/messages/market_data/request_cancel_data_test.exs (hunk 5 above)
Comment (ooddaa) [reply test/ib_ex/client/messages/market_data/request_cancel_data_test.exs:11]: Because this the only way I can reliably pass the `request_id` of the `%MarketData.RequestData{}` to be cancelled. If I don't do that, the `reverse_lookup/2` will match the first registered id by pid, which might not be (and in my case never was) the correct `request_id`. 

So I'd send 
```zsh
19:31:17.014 [info] --> MarketData.RequestData{
  request_id: 2,
  contract: 0, NVDA, OPT, 20241220, 150.0, C, 100, SMART, , USD, , ,
  tick_list: ,
  snapshot: false,
  regulatory_snapshot: false
}
```
Then when I'm trying to cancel this request from the same shell session via 
```zsh
{:ok, msg} = MarketData.RequestMarketData.new()
Client.send_request(pid, msg)
```
I get 
```zsh
19:31:25.165 [info] <-- %Error{id: 1, code: 300, version: 2, message: Can't find EId with tickerId:1}
```
It's trying to cancel `request_id: 1` instead of `2`

Comment (ooddaa) [reply test/ib_ex/client/messages/market_data/request_cancel_data_test.exs]: as per the slack discussion, gonna revert back these changes

File: lib/ib_ex/client/messages/market_data/request_cancel_data.ex (hunk 4 above)
Comment (ooddaa) [reply lib/ib_ex/client/messages/market_data/request_cancel_data.ex]: reverted
